enhancements/machine-api/control-plane-machineset-vertical-scaling.md

```diff
@@ -0,0 +1,348 @@
+
+### User Stories
+
+* As an OpenShift administrator, I want to know that the platform will always
+  have enough resources available to it's control plane nodes without manual
+  intervention.
+* As an OpenShift administrator, I want to know that the platform can reduce
+  spending on control plane nodes if they are oversized for the current cluster
+  load.
+* As an OpenShift administrator, I want to be able to control when sizing up and
+  down occurs, to adjust it to workloads on my cluster that might be burstable
+  to prevent the automatic scaling from occurring to often.
+
+### Goals
+
+* Allow automatic vertical scaling of control plane nodes.
+* Allow users to configure the sizes that can be chosen for automatic scaling.
+* Allow users to configure the thresholds when scaling will occur for CPU and
+  memory usage.
+* Allow users to configure how often checks for automatic scaling should be run.
+* Allow users to configure how long after a scale up or scale down, no more
+  scaling should be performed.
+* Allow users to completely disable scale up.
+* Allow users to completely disable scale down.
+
+### Non-Goals
+
+* Horizontal scaling of control plane nodes
+
+## Proposal
+
+New configuration and a reconcile loop for the
+`control-plane-machine-set-operator` will be introduced that will monitor
+resource usage of the control plane nodes and automatically modify the
+`ControlPlaneMachineSet` if scaling the nodes is determined to be required.
+
+The new configuration will specify if scaling up and down is enabled and allow
+configuring when to trigger scaling. It will also allow configuring which
+machine types are to be used for scaling, putting an upper and lower limit on
+possible automatic scaling.
+
+### Workflow Description
+
+1. The OpenShift administrator installs the
+   `control-plane-machine-set-operator`.
+2. The OpenShift adminstrator creates a valid `control-plane-machine-set`
+   `configmap` in the `openshift-machine-api` namespace (or the respective
+   namespace `control-plane-machineset-operator` is running in), to configure
+   the automatic vertical scaling.
+
+### API Extensions
+
+This proposal does not require API extensions.
+
+### Implementation Details/Notes/Constraints [optional]
+
+The implementation for this feature, requires a data source to gather current
+CPU and memory usage for nodes, as well as available resources on the nodes.
+
+To gather available resource information the default
+[Node](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.27/#node-v1-core)
+API endpoint can be used.
+
+Using the [horizontal pod
+autoscaler](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/)
+as a reference, they are using the [metrics
+API](https://kubernetes.io/docs/tasks/debug/debug-cluster/resource-metrics-pipeline/)
+to gather performance statistics for nodes. 
+
+Using the metrics API has the least requirements as it does not expect a running
```

Comment:
And this is available in any openshift cluster without any additional configuration?

Reply:
AFAIK in OpenShift this should be the case, due to our preconfigured monitoring stack that provices the APIService:
```
oc get apiservice | grep metrics
v1beta1.metrics.k8s.io                          openshift-monitoring/prometheus-adapter                      True        38m
```